### DISCUSSION TITLE
[infinity] errors and exceptions while using infinity samples

### DISCUSSION THREAD

**Discussion by nainamd77** [infiniflow/infinity#602] [Q&A] [answered]
Hi, I have had tried samples provided on infinity db. While testing with query builder, it thrown exception as follows 
**'InfinityThriftQueryBuilder' object is not callable**. Also I have tried some other features based on sample code and got exceptions too. I have added the sample program and the exceptions thrown.
**Env: Python Code, VSCode Editor, Ubuntu 22.0**

_query_br = table_obj.query_builder()_
**'InfinityThriftQueryBuilder' object is not callable**

_qb_result1 = table_obj.output(["num", "body"]).match("body","bloom","float","1").to_pl()_
**InfinityThriftQueryBuilder.match() takes from 3 to 4 positional arguments but 5 were given**

_qb_result1 = table_obj.output(["num", "body"]).match("body","bloom","topn=1").to_pl()_
**RemoteTable.match() missing 1 required positional argument: 'topn'**

_qb_result1 = table_obj.output(["num", "body"]).fusion('rrf').to_pl()_
**'RemoteTable' object has no attribute 'fusion'.**

I have added full program file in that
[inf_main_t.txt](https://github.com/infiniflow/infinity/files/14328490/inf_main_t.txt)

Please guide us to fix and test above. Thanks

**Comment by loloxwg** [comment]
Please provide the infinity-sdk version you are using now.

**Reply by nainamd77** [reply]
infinity-sdk version 0.1.0.dev4

**Reply by loloxwg** [reply]
Update to 0.1.0-dev5, try again later

**Reply by yuzhichang** [reply]
I have reproduced the issue with 0.1.0-dev5. Try fixing...

**Reply by yuzhichang** [reply]
There's bug inside infinify-sdk 0.1.0-dev5. I've released the fix in infinify-sdk 0.1.0-dev6. You can reinstall the sdk like this:
```
$ pip3 uninstall infinity-sdk
$ pip3 install infinity-sdk
```

And there's bug inside your python script. Here's the updated script:
```
import json
import infinity
import infinity.index as index
from infinity.common import REMOTE_HOST
import pandas as pds


def main():
    try:
        infinity_obj = infinity.connect(REMOTE_HOST)
        db = infinity_obj.get_database("default")
        # Drop my_table if it already exists
        db.drop_table("my_table", if_exists=True)
        # Create a table named "my_table"
        table = db.create_table("my_table", {
                                "num": "integer", "body": "varchar", "vec": "vector, 4, float"}, None)
        table.insert(
            [{"num": 1, "body": "unnecessary and harmful", "vec": [1.0, 1.2, 0.8, 0.9]}])
        table.insert(
            [{"num": 2, "body": "Office for Harmful Blooms", "vec": [4.0, 4.2, 4.3, 4.5]}])

        # `create_index()` is required before match() or fusion()
        res = table.create_index("my_index",
                                 [index.IndexInfo("body",
                                                  index.IndexType.FullText,
                                                  [index.InitParameter("ANALYZER", "segmentation")]),
                                  ], None)
        assert res.success

        res = table.output(["num", "body"]).knn(
            "vec", [3.0, 2.8, 2.7, 3.1], "float", "ip", 2).to_pl()

        pds_df = pds.DataFrame(res)
        json_data = pds_df.to_json()

        table_obj = db.get_table("my_table")
        # query_b = table_obj.query_builder()
        print("------knn-------")
        qb_result = table_obj.output(["num", "body"]).knn(
            "vec", [3.0, 2.8, 2.7, 3.1], "float", "ip", 3).to_pl()

        print("------match-------")
        qb_result1 = table_obj.match(
            "body", "bloom", "topn=1").output(["num", "body"]).to_pl()

        print("------knn+match-------")
        qb_result2 = table_obj.output(["num", "body"]).knn(
            "vec", [3.0, 2.8, 2.7, 3.1], "float", "ip", 3).match(
            "body", "bloom", "topn=1").fusion('rrf').to_pl()

        print("------json-------")
        print(json_data)
        print("------tabular - querybuilder-------")
        print(qb_result)
        print(qb_result1)
        print(qb_result2)

    except Exception as e:
        print(str(e))

    # Print the JSON data
    # print("------json-------")
    # print(json_data)
    # print("------tabular - querybuilder-------")
    # print(qb_result)


if __name__ == '__main__':
    main()

```

**Reply by nainamd77** [reply]
ok thanks, i will try above and update

**Comment by nainamd77** [comment]
Hi, 
Now I have installed _infinify-sdk 0.1.0-dev6_ as suggested,
but still I got exception while calling query_builder object as below
"'InfinityThriftQueryBuilder' object is not callable" @ query_b = table_obj.query_builder().
 Other issues I have encountered in Match, Fusion functionalities fixed once index created.

_Logs_
[infi_log.txt](https://github.com/infiniflow/infinity/files/14373363/infi_log.txt)

**Another issue: Infinity crashed**
While doing testing suddenly infinity stopped and an uncaught exception thrown in application as belows

TSocket read 0 bytes
Exception ignored in: <function RemoteThriftInfinityConnection.__del__ at 0x7f4eee197760>
Traceback (most recent call last):
  File "/home/barrla/.local/lib/python3.10/site-packages/infinity/remote_thrift/infinity.py", line 33, in __del__
    self.disconnect()
  File "/home/barrla/.local/lib/python3.10/site-packages/infinity/remote_thrift/infinity.py", line 75, in disconnect
    res = self._client.disconnect()
  File "/home/barrla/.local/lib/python3.10/site-packages/infinity/remote_thrift/client.py", line 155, in disconnect
    res = self.client.Disconnect(CommonRequest(session_id=self.session_id))
  File "/home/barrla/.local/lib/python3.10/site-packages/infinity/remote_thrift/infinity_thrift_rpc/InfinityService.py", line 242, in Disconnect
    return self.recv_Disconnect()
  File "/home/barrla/.local/lib/python3.10/site-packages/infinity/remote_thrift/infinity_thrift_rpc/InfinityService.py", line 254, in recv_Disconnect
    (fname, mtype, rseqid) = iprot.readMessageBegin()
  File "/home/barrla/.local/lib/python3.10/site-packages/thrift/protocol/TBinaryProtocol.py", line 134, in readMessageBegin
    sz = self.readI32()
  File "/home/barrla/.local/lib/python3.10/site-packages/thrift/protocol/TBinaryProtocol.py", line 217, in readI32
    buff = self.trans.readAll(4)
  File "/home/barrla/.local/lib/python3.10/site-packages/thrift/transport/TTransport.py", line 62, in readAll
    chunk = self.read(sz - have)
  File "/home/barrla/.local/lib/python3.10/site-packages/thrift/transport/TTransport.py", line 164, in read
    self.__rbuf = BufferIO(self.__trans.read(max(sz, self.__rbuf_size)))
  File "/home/barrla/.local/lib/python3.10/site-packages/thrift/transport/TSocket.py", line 166, in read
    raise TTransportException(type=TTransportException.END_OF_FILE,
thrift.transport.TTransport.TTransportException: TSocket read 0 bytes

_Log with crash info_
[Exception-details.txt](https://github.com/infiniflow/infinity/files/14373419/Exception-details.txt)

_Note:_
This might happened as I was wrongly tried to create another index on same table and same column which already have an index created. But instead of throwing normal exception, it crashed. 
Once I fixed that index creation part, it is working without any issue(crash) as of now.

**Reply by JinHai-CN** [reply]
1. By default, the index isn't created. And MATCH and FUSION function is only used by full-text search, which needs the index. We will improve this behavior in the future.

2. As for query_builder issues, we will re-check the Python-SDK and come back to you later.

3. As for crash, we will try to reproduce it according to your description.

4. We suggest you to update the infinity binary to the latest version. We fixed several bugs in these days. From the Exception-details.txt, I can find you are using infinity built in 20240212.
`Infinity, version: 0.1.0 build on 2024-02-12 00:50.22 with Release mode from branch: main, commit-id: 0b212a1ad157a3a0705cac03690152ccbd38e5f4`

**Reply by JinHai-CN** [reply, marked as answer]
If index isn't built, calling MATCH and FUSION would introduce crash. This might be your case, and it is fixed by https://github.com/infiniflow/infinity/pull/651.

**Reply by chrysanthemum-boy** [reply]
There was an error in the `pysdk_api_reference.md`, and we have updated it, and query_builder() is not supported at this time.

**Comment by nainamd77** [comment]
hi,
encountering a strange problem that  _Match function not working_ as expected below the details of issue,

First I created one table along with index and loaded with data, then tried Match function and got expected result.
Then I added another table  along with index and loaded with data , then tried Match function but it returns empty result. Even I tried with query_builder but still same problem.
Attached the code file. 
Please check and help us to fix.
[test_code_python.txt](https://github.com/infiniflow/infinity/files/14468905/test_code_python.txt)

Got _**Infinity crashed**_ when I tried to create index with a unavailable column name.
_build-version_
infinity-sdk 0.1.0.dev7
infinity 0.1.0 build on 2024-02-26 00:50

**Reply by yuzhichang** [reply]
line 69-75, you create two fulltext index. It's not supported yet. Currently we assmue only one fulltext index on a table.
line 81-129, you insert lines after `CREATE INDEX`. It's not supported yet. Currently we require inserting lines before `CREATE INDEX` .

Release 0.2.0  (hopefully April) will relax the above restrictions.

**Reply by nainamd77** [reply]
Thanks for replying, I will modify accordingly and test it.
